pkg/cmd/certregenerationcontroller/certregenerationcontroller.go

```diff
@@ -0,0 +1,208 @@
+	}
+
+	o.kubeClient, err = kubernetes.NewForConfig(o.restConfig)
+	if err != nil {
+		return fmt.Errorf("can't build kubernetes clientset: %v", err)
+	}
+
+	o.configClient, err = configeversionedclient.NewForConfig(o.restConfig)
+	if err != nil {
+		return fmt.Errorf("failed to create config client: %v", err)
+	}
+
+	return nil
+}
+
+func (o *Options) Run() error {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	signalHandler := server.SetupSignalHandler()
+	go func() {
+		<-signalHandler
+		cancel()
+	}()
+
+	configInformers := configexternalinformers.NewSharedInformerFactory(o.configClient, 10*time.Minute)
+
+	kubeAPIServerInformersForNamespaces := v1helpers.NewKubeInformersForNamespaces(
+		o.kubeClient,
+		operatorclient.GlobalMachineSpecifiedConfigNamespace,
+		operatorclient.GlobalUserSpecifiedConfigNamespace,
+		operatorclient.OperatorNamespace,
+		operatorclient.TargetNamespace,
+	)
+
+	operatorClient, dynamicInformers, err := genericoperatorclient.NewStaticPodOperatorClient(o.restConfig, operatorv1.GroupVersion.WithResource("kubeapiservers"))
+	if err != nil {
+		return err
+	}
+
+	eventRecorder := events.NewKubeRecorder(o.kubeClient.CoreV1().Events(""), "openshift-kube-apiserver-operator", &corev1.ObjectReference{
+		APIVersion: "v1",
+		Kind:       "namespace",
+		Name:       "openshift-kube-apiserver-operator",
+		Namespace:  "openshift-kube-apiserver-operator",
+	}) // this is a fake object-reference that should hopefully place us in the correct namespace
+
+	certRotationScale, err := certrotation.GetCertRotationScale(o.kubeClient, operatorclient.GlobalUserSpecifiedConfigNamespace)
+	if err != nil {
+		return err
+	}
+
+	kubeAPIServerCertRotationController, err := certrotationcontroller.NewCertRotationController(
+		o.kubeClient,
+		operatorClient,
+		configInformers,
+		kubeAPIServerInformersForNamespaces,
+		eventRecorder.WithComponentSuffix("cert-rotation-controller"),
+		certRotationScale,
+		true,
+	)
+	if err != nil {
+		return err
+	}
+
+	configMapLister := kubeAPIServerInformersForNamespaces.ConfigMapLister()
+
+	// We can't start informers until after the resources have been requested. Now is the time.
+	configInformers.Start(ctx.Done())
+	kubeAPIServerInformersForNamespaces.Start(ctx.Done())
+	dynamicInformers.Start(ctx.Done())
+
+	// wait until the controllers are ready
+	kubeAPIServerCertRotationController.Run(ctx, o.Workers)
+
+	if !cache.WaitForCacheSync(
+		ctx.Done(),
+		kubeAPIServerInformersForNamespaces.InformersFor(operatorclient.GlobalUserSpecifiedConfigNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+		kubeAPIServerInformersForNamespaces.InformersFor(operatorclient.OperatorNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+		kubeAPIServerInformersForNamespaces.InformersFor(operatorclient.TargetNamespace).Core().V1().ConfigMaps().Informer().HasSynced,
+	) {
+		return fmt.Errorf("caches did not sync")
+	}
+	go wait.UntilWithContext(ctx, func(ctx context.Context) {
```

Comment:
@deads2k How real we want to get w.r.t. to these control loops? Running the full enclosing controller with queue seemed pretty heavy weight and it could steal work from the operator, this steals just a small loop which runs in parallel.

Reply:
You should create a different file with a `NewFooController` and a `Run(context) function` for its structural value.  I don't care if it has a queue or not or triggering functions, but the encapsulation of "this is my input, this is what I use for state, this is how I start" makes maintenance and testing easier.

Reply:
done
(I wired the queue as well when I was making an object for it.)